homeassistant/components/mqtt/client.py

```diff
@@ -521,16 +629,26 @@ async def async_connect(self) -> None:
                 "Failed to connect to MQTT server: %s", mqtt.error_string(result)
             )
 
-        self._mqttc.loop_start()
+        self._reconnect_task = self.config_entry.async_create_background_task(
+            self.hass, self._reconnect_loop(), "mqtt reconnect loop"
+        )
+
+    async def _reconnect_loop(self) -> None:
+        """Reconnect to the MQTT server."""
+        while True:
+            if not self.connected:
+                try:
+                    await self.hass.async_add_executor_job(self._mqttc.reconnect)
+                except OSError as err:
+                    _LOGGER.debug(
+                        "Error re-connecting to MQTT server due to exception: %s", err
+                    )
+
+            await asyncio.sleep(RECONNECT_INTERVAL_SECONDS)
```

Comment:
This could probably be a bit smarter and reconnect right away after disconnect?

Reply:
We could also make it a bit smarter to reconnect right away on discovery as well

Reply:
Maybe we need exp backoff as well?